crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -83,69 +215,105 @@ impl<'db> UseDefMapBuilder<'db> {
+    /// path to get here. The new visible-definitions state for each symbol should include
+    /// definitions from both the prior state and the snapshot.
     pub(super) fn merge(&mut self, state: &FlowSnapshot) {
+        // The tricky thing about merging two Ranges pointing into `all_definitions` is that if the
+        // two Ranges aren't already adjacent in `all_definitions`, we will have to copy at least
+        // one or the other of the ranges to the end of `all_definitions` so as to make them
+        // adjacent. We can't ever move things around in `all_definitions` because previously
+        // recorded uses may still have ranges pointing to any part of it; all we can do is append.
+        // It's possible we may end up with some old entries in `all_definitions` that nobody is
+        // pointing to, but that's OK.
+
         for (symbol_id, to_merge) in state.definitions_by_symbol.iter_enumerated() {
             let current = &mut self.definitions_by_symbol[symbol_id];
-            // if the symbol can be unbound in either predecessor, it can be unbound
+
+            // If the symbol can be unbound in either predecessor, it can be unbound post-merge.
             current.may_be_unbound |= to_merge.may_be_unbound;
-            // merge the definition ranges
-            if current.definitions == to_merge.definitions {
-                // ranges already identical, nothing to do!
-            } else if current.definitions.end == to_merge.definitions.start {
-                // ranges adjacent (current first), just merge them
-                current.definitions = (current.definitions.start)..(to_merge.definitions.end);
-            } else if current.definitions.start == to_merge.definitions.end {
-                // ranges adjacent (to_merge first), just merge them
-                current.definitions = (to_merge.definitions.start)..(current.definitions.end);
-            } else if current.definitions.end == self.all_definitions.len() {
-                // ranges not adjacent but current is at end, copy only to_merge
+
+            // Merge the definition ranges.
+            if current.definitions_range == to_merge.definitions_range {
+                // Ranges already identical, nothing to do!
+            } else if current.definitions_range.end == to_merge.definitions_range.start {
+                // Ranges are adjacent (`current` first), just merge them into one range.
+                current.definitions_range =
+                    (current.definitions_range.start)..(to_merge.definitions_range.end);
+            } else if current.definitions_range.start == to_merge.definitions_range.end {
+                // Ranges are adjacent (`to_merge` first), just merge them into one range.
+                current.definitions_range =
+                    (to_merge.definitions_range.start)..(current.definitions_range.end);
+            } else if current.definitions_range.end == self.all_definitions.len() {
```

Comment:
I think it could be helpful to add a short code example of when each branch is possible. 

To me it's not clear when this branch can be taken. My intuition is that `current.definitions` is always shorter than the `state` we merge in? Is this to handle the case when merging multiple branches? Then each branch could add new definitions.

Reply:
The choice of these branches isn't impacted by which definition range is shorter or longer. It's about their relative position. If they are adjacent, then we can easily merge them into one range. If they are not adjacent but one is already at the end of `all_definitions`, we can just copy the other one to the end, and now they are adjacent. If they are not adjacent and neither one is at the end, we have to copy both to the end to make them adjacent.

I don't think there are any guarantees we can make about which of these cases are possible that couldn't easily be violated when we add support for future branching constructs (or even change the implementation slightly for one of the current branching constructs.) It all depends on which order we merge things in.

In https://github.com/astral-sh/ruff/pull/12373 for instance I make some efficiency improvements to if-statement visiting which change the order of restores and merges (in order to do fewer overall) and thus change which of these branches are taken for which specific code examples.

In that PR I also expand code coverage of inference examples so that it actually covers all of these branches, at least for now.

What might be ideal would be to have unit tests of this logic, independent of code examples, to verify at a lower level that all the cases here work correctly. But that seems moderately painful to do, since it would require manually creating a bunch of `Definition`s. I might play around with just how painful it is.

Reply:
(But I don't think comments in this code giving code samples per-branch would be helpful, since it depends on exactly how the visit is implemented, which is external to this code.)